src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs

```diff
@@ -0,0 +1,246 @@
+
+#nullable enable
+
+using System;
+using System.Collections;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    /// <summary>
+    /// Test binding of the target-typed conditional (aka ternary) operator.
+    /// </summary>
+    public class TargetTypedConditionalOperatorTests : CSharpTestBase
+    {
+        [Fact]
+        public void TestImplicitConversions_Good()
+        {
+            // Implicit constant expression conversions
+            TestConditional("b ? 1 : 2", "System.Int16", "System.Int32",
+                // (6,26): error CS0266: Cannot implicitly convert type 'int' to 'short'. An explicit conversion exists (are you missing a cast?)
+                //         System.Int16 t = b ? 1 : 2;
+                Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "b ? 1 : 2").WithArguments("int", "short").WithLocation(6, 26)
+                );
+
+            // Implicit reference conversions
+            TestConditional("b ? GetB() : GetC()", "A", null);
+
+            // Implicit numeric conversions
+            TestConditional("b ? GetUInt() : GetInt()", "System.Int64", null);
+
+            // Implicit enumeration conversions
+            TestConditional("b ? 0 : 0", "color", "System.Int32",
+                // (6,19): error CS0266: Cannot implicitly convert type 'int' to 'color'. An explicit conversion exists (are you missing a cast?)
+                //         color t = b ? 0 : 0;
+                Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "b ? 0 : 0").WithArguments("int", "color").WithLocation(6, 19)
+                );
+
+            // Implicit interpolated string conversions
+            TestConditional(@"b ? $""x"" : $""x""", "System.FormattableString", "System.String",
+                // (6,38): error CS0029: Cannot implicitly convert type 'string' to 'System.FormattableString'
+                //         System.FormattableString t = b ? $"x" : $"x";
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, @"b ? $""x"" : $""x""").WithArguments("string", "System.FormattableString").WithLocation(6, 38)
+                );
+
+            // Implicit nullable conversions
+            // Null literal conversions
+            TestConditional("b ? 1 : null", "System.Int64?", null);
+
+            // Boxing conversions
+            TestConditional("b ? GetUInt() : GetInt()", "System.IComparable", null);
+
+            // User - defined implicit conversions
+            TestConditional("b ? GetB() : GetC()", "X", null);
+
+            // Anonymous function conversions
+            TestConditional("b ? a=>a : b=>b", "Del", null);
+
+            // Method group conversions
+            TestConditional("b ? M1 : M2", "Del", null);
+        }
```

Comment:
Consider adding success case tests for implicit dynamic conversion (operands are completely unrelated)

Reply:
The *implicit dynamic conversion* is from an expression of type `dynamic` to any type.  If one of the arguments were of type `dynamic`, then it would not be target-typed.  If the target were of type `dynamic`, then that is the same as the target type being of type `object`.

---
In reply to: [410449513](https://github.com/dotnet/roslyn/pull/40604#discussion_r410449513) [](ancestors = 410449513)